### PR TITLE
Auto run `dcos auth login` when token expired

### PR DESCRIPTION
**Pull request by msabramo**
if `core.dcos_auto_login` is set.

```
$ grep auto ~/.dcos/dcos.toml
dcos_auto_login = true

$ dcos node
If your browser didn't open, please go to the following link:

    http://<redacted>/login?redirect_uri=urn:ietf:wg:oauth:2.0:oob

Enter OpenID Connect ID Token: <paste token>
Login successful!
  HOSTNAME         IP                          ID
10.10.12.174  10.10.12.174  74df965e-a46e-46a9-8d2a-639db0af0e25-S0
10.10.12.183  10.10.12.183  74df965e-a46e-46a9-8d2a-639db0af0e25-S4
10.10.12.29   10.10.12.29   74df965e-a46e-46a9-8d2a-639db0af0e25-S2
10.10.3.201   10.10.3.201   74df965e-a46e-46a9-8d2a-639db0af0e25-S3
```

### CODE REVIEW

File: dcos/http.py [hunk 1]
```diff
@@ -149,6 +153,7 @@ def request(method,
 
     toml_config = config.get_config()
     auth_token = config.get_config_val("core.dcos_acs_token", toml_config)
+    auth_token_auto_login = config.get_config_val("core.dcos_auto_login", toml_config)
```
Comment (tamarrow) [inline dcos/http.py]: lets call this `core.prompt_login`. We should add it to the schema for "core": dcos/data/config-schema/core.json

File: dcos/http.py [hunk 2]
```diff
@@ -169,6 +174,10 @@ def request(method,
         if auth_token is not None:
             msg = ("Your core.dcos_acs_token is invalid. "
                    "Please run: `dcos auth login`")
+            if auth_token_auto_login:
+                Subproc().call(
```
Comment (tamarrow) [inline dcos/http.py]: this should just call the auth module, and not fork.

File: dcos/http.py (hunk 1 above)
Comment (msabramo) [inline dcos/http.py]: 57e6ed1475f8979a9c793720db8011632f8d75e4

File: dcos/http.py (hunk 2 above)
Comment (msabramo) [inline dcos/http.py]: 6017abfe8969df996cd53379bb1a8380796c0505

**Comment by msabramo** [comment]
@tamarrow: Addressed both comments. Thanks!

**Comment by msabramo** [comment]
@tamarrow: Any other changes needed here?

File: dcos/http.py [hunk 3]
```diff
@@ -167,9 +173,13 @@ def request(method,
...
-            msg = ("Your core.dcos_acs_token is invalid. "
-                   "Please run: `dcos auth login`")
-            raise DCOSAuthenticationException(msg)
+            if prompt_login:
+                msg = ("Your core.dcos_acs_token is invalid. "
+                       "Please run: `dcos auth login`")
+                raise DCOSAuthenticationException(msg)
+            else:
+                header_challenge_auth(dcos_url.geturl())
+                os.execl(sys.argv[0], *sys.argv)
```
Comment (tamarrow) [inline dcos/http.py]: we don't need this line anymore

Comment (msabramo) [inline dcos/http.py]: Did you try it without this line? I'm on my phone so I can't check.

Comment (msabramo) [inline dcos/http.py]: I just tried it and it worked (awesome!) so I updated it. 17e4fc3.

File: dcos/http.py [hunk 4]
```diff
@@ -167,9 +173,12 @@ def request(method,
         return response
     elif response.status_code == 401:
         if auth_token is not None:
-            msg = ("Your core.dcos_acs_token is invalid. "
-                   "Please run: `dcos auth login`")
-            raise DCOSAuthenticationException(msg)
+            if prompt_login:
```
Comment (tamarrow) [inline dcos/http.py]: you have the if/else cases switched. `if prompt_login` we want to call `header_challenge_auth`

Comment (tamarrow) [inline dcos/http.py]: we should call the auth method whether or not there is an auth token. The original condition is for the message we show to the user.

File: dcos/http.py [hunk 5]
```diff
@@ -167,9 +173,12 @@ def request(method,
...
+                msg = ("Your core.dcos_acs_token is invalid. "
+                       "Please run: `dcos auth login`")
+                raise DCOSAuthenticationException(msg)
+            else:
+                header_challenge_auth(dcos_url.geturl())
```
Comment (tamarrow) [inline dcos/http.py]: we need to call `request` with the same parameters after this call. The reason it works for some commands is that we do checks before the actual call, but this won't be true for all commands.

File: dcos/http.py (hunk 4 above)
Comment (msabramo) [inline dcos/http.py]: OK, updated.

File: dcos/http.py (hunk 5 above)
Comment (msabramo) [inline dcos/http.py]: OK, updated.

File: dcos/http.py [hunk 6]
```diff
@@ -166,12 +172,24 @@ def request(method,
...
-                   "Please run: `dcos auth login`")
-            raise DCOSAuthenticationException(msg)
+        if prompt_login:
+            header_challenge_auth(dcos_url.geturl())
+            auth_token = config.get_config_val("core.dcos_acs_token", toml_config)
+            if auth_token and scheme_eq and netloc_eq:
+                auth = DCOSAcsAuth(auth_token)
+            else:
+                auth = None
+            response = _request(method, url, is_success, timeout,
```
Comment (tamarrow) [inline dcos/http.py]: I'm not sure this is what we want. We are missing several of the status code errors and duplicating code. Why don't we just call `request` again after `header_challenge_auth`. If that method succeeds it means that we auth-ed correctly and thus won't get a 401 (and therefore a recursive loop). We could also rerun `_request` right after the first call to `_request` if we get a 401.

Comment (msabramo) [inline dcos/http.py]: Ah sorry I misunderstood your comment before. You said `request` and I didn't notice the function was called `request` and thought you were referring to `_request`.

f673b3e

**Comment by msabramo** [comment]
Oy. I wish we could get rid of these CCM errors. I don't know what CCM is, but it seems to fail a lot.

screen shot 2017-03-14 at 7 11 39 pm

**Comment by tamarrow** [comment]
Ha yea... it stands for Cluster Create Manager. It's a wrapper around aws, but I hear it should get re-written soon.

**Comment by tamarrow** [comment]
ok to test

**Comment by msabramo** [comment]
[auth_token_auto_login f3881bc] Remove unnecessary imports
 1 file changed, 4 deletions(-)

**Comment by msabramo** [comment]
screen shot 2017-03-15 at 3 33 24 pm

Huh?

**Comment by msabramo** [comment]
Fixed:

```
dcos/http.py:177:80: E501 line too long (81 > 79 characters)
```

with 059ee29.

**Comment by tamarrow** [comment]
jwt is already installed in setup

**Comment by msabramo** [comment]
but I reproduced the error locally:

```
$ git checkout 059ee29d90a62c03f413f671efd222669b5dcfd4
...
HEAD is now at 059ee29... Fix line too long error

$ tox -e py35-unit
GLOB sdist-make: /Users/abramowi/dev/git-repos/dcos-cli/setup.py
py35-unit inst-nodeps: /Users/abramowi/dev/git-repos/dcos-cli/.tox/dist/dcos-SNAPSHOT.zip
py35-unit installed: appdirs==1.4.3,coverage==4.3.4,dcos===SNAPSHOT,jsonschema==2.6.0,mock==2.0.0,packaging==16.8,pager==3.3,pbr==2.0.0,prettytable==0.7.2,py==1.4.32,Pygments==2.2.0,pyparsing==2.2.0,pytest==3.0.7,pytest-cov==2.4.0,requests==2.13.0,six==1.10.0,sseclient==0.0.14,teamcity-messages==1.21,toml==0.9.2
...
=================================================================================================== ERRORS ===================================================================================================
__________________________________________________________________________________ ERROR collecting tests/test_marathon.py ___________________________________________________________________________________
ImportError while importing test module '/Users/abramowi/dev/git-repos/dcos-cli/tests/test_marathon.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
.tox/py35-unit/lib/python3.5/site-packages/_pytest/python.py:418: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
.tox/py35-unit/lib/python3.5/site-packages/py/_path/local.py:662: in pyimport
    __import__(modname)
.tox/py35-unit/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:216: in load_module
    py.builtin.exec_(co, mod.__dict__)
tests/test_marathon.py:9: in <module>
    from dcos import http, marathon, rpcclient
.tox/py35-unit/lib/python3.5/site-packages/dcos/http.py:7: in <module>
    from dcos.auth import header_challenge_auth
.tox/py35-unit/lib/python3.5/site-packages/dcos/auth.py:7: in <module>
    import jwt
E   ImportError: No module named 'jwt'
_______________________________________________________________________________ ERROR collecting tests/test_packagemanager.py ________________________________________________________________________________
ImportError while importing test module '/Users/abramowi/dev/git-repos/dcos-cli/tests/test_packagemanager.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
.tox/py35-unit/lib/python3.5/site-packages/_pytest/python.py:418: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
.tox/py35-unit/lib/python3.5/site-packages/py/_path/local.py:662: in pyimport
    __import__(modname)
.tox/py35-unit/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:216: in load_module
    py.builtin.exec_(co, mod.__dict__)
tests/test_packagemanager.py:5: in <module>
    from dcos import packagemanager
.tox/py35-unit/lib/python3.5/site-packages/dcos/packagemanager.py:8: in <module>
    from dcos import cosmos, emitting, util
.tox/py35-unit/lib/python3.5/site-packages/dcos/cosmos.py:3: in <module>
    from dcos import config, http, util
.tox/py35-unit/lib/python3.5/site-packages/dcos/http.py:7: in <module>
    from dcos.auth import header_challenge_auth
.tox/py35-unit/lib/python3.5/site-packages/dcos/auth.py:7: in <module>
    import jwt
E   ImportError: No module named 'jwt'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================================================================== 2 error in 1.01 seconds ===========================================================================================
ERROR: InvocationError: '/Users/abramowi/dev/git-repos/dcos-cli/.tox/py35-unit/bin/py.test -p no:cacheprovider -vv --cov /Users/abramowi/dev/git-repos/dcos-cli/.tox/py35-unit/lib/python3.5/site-packages/dcos tests'
__________________________________________________________________________________________________ summary ___________________________________________________________________________________________________
ERROR:   py35-unit: commands failed
```

**Comment by msabramo** [comment]
```
$ grep -c 'jwt' setup.py requirements.txt
setup.py:0
requirements.txt:0
```

**Comment by tamarrow** [comment]
https://github.com/dcos/dcos-cli/blob/master/cli/setup.py#L69

**Comment by msabramo** [comment]
Oh I see.

But:

```
$ grep -i -c 'jwt' setup.py requirements.txt
setup.py:0
requirements.txt:0

abramowi at abramowi-macOS in ~/dev/git-repos/dcos-cli (059ee29...●)
$ ag jwt dcos/auth.py
7:import jwt
21:       'acsjwt' (DC/OS acs auth), 'oauthjwt' (DC/OS acs oauth), or
33:                       if auth_type.rstrip().lower().startswith("acsjwt") or
34:                       auth_type.rstrip().lower().startswith("oauthjwt")),
42:                   "'{}', DC/OS only supports ['oauthjwt', 'acsjwt']".format(
89:    jwt token encoded with private key).
257:        'token': jwt.encode(
323:            if auth_scheme == "oauthjwt":
325:            # auth_scheme == "acsjwt"
```

so `dcos/cli` has the dependency on `jwt`, but I think `dcos` should have it, since it's used in `dcos/auth.py`.

In fact, it looks like it's only used in `dcos/auth.py` and not used at all in `dcos/cli`:

```
$ ag --ignore=tests --ignore=env --ignore=venv jwt --ignore='*.egg-info'
cli/setup.py
69:        'PyJWT==1.4.2',

dcos/auth.py
7:import jwt
21:       'acsjwt' (DC/OS acs auth), 'oauthjwt' (DC/OS acs oauth), or
33:                       if auth_type.rstrip().lower().startswith("acsjwt") or
34:                       auth_type.rstrip().lower().startswith("oauthjwt")),
42:                   "'{}', DC/OS only supports ['oauthjwt', 'acsjwt']".format(
89:    jwt token encoded with private key).
257:        'token': jwt.encode(
323:            if auth_scheme == "oauthjwt":
325:            # auth_scheme == "acsjwt"
```

so I think the dependency should be moved. See https://github.com/dcos/dcos-cli/pull/937

**Comment by msabramo** [comment]
Another CCM failure.

**Comment by msabramo** [comment]
retest

**Comment by msabramo** [comment]
Hmmm I thought my other change would fix the CI failure. Any other ideas?

File: cli/tests/data/dcos.toml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 [core]
 dcos_url = "http://dcos.snakeoil.mesosphere.com"
 ssl_verify = "false"
-timeout = 5
 reporting = false
+timeout = 5
```
Comment (tamarrow) [inline cli/tests/data/dcos.toml]: remove this file and lets get this merged :)

File: dcos/http.py [hunk 8]
```diff
@@ -169,6 +168,11 @@ def request(method,
         return response
     elif response.status_code == 401:
         if prompt_login:
+            # I don't like having imports that aren't at the top level, but
+            # this is to resolve a circular import issue between dcos.http and
+            # dcos.auth
+            from dcos.auth import header_challenge_auth
```
Comment (tamarrow) [inline dcos/http.py]: :( make sense

File: cli/tests/data/dcos.toml (hunk 7 above)
Comment (msabramo) [inline cli/tests/data/dcos.toml]: d58336b

👍

**Comment by msabramo** [comment]
For those watching at home who didn't see the commit 649bf1a, it turned out that the test failure was because of a circular import and it's fixed by 649bf1a.

I also just removed two changes that aren't necessary in d58336b and 7b06b03.

**Comment by msabramo** [comment]
CCM cluster error

**Comment by msabramo** [comment]
Ooh CI on this one finally went green! 🎉

**Comment by msabramo** [comment]
🎉

**Pull request merged**